epyqlib/qscale.py

```diff
@@ -238,154 +283,324 @@ def paintEvent(self, paintEvent):
+                            else:
+                                # Color scale orientation changed if flipped.
+                                # Keep in mind, it isn't a PERFECT flip.
+                                qpp.arcMoveTo(x, -y - d, d, d, rangeAngleStart)
+                                qpp.arcTo(x, -y - d, d, d, rangeAngleStart, 
+                                          rangeAngleSpan)
+                            outer = QtGui.QPainterPath()
+                            r = radius - 0.6 * scaleWidth
+                            d = 2 * r
+                            x = center.x() - r
+                            y = center.y() - r
+
+                            if not (vertical and self.vertically_flipped):
+                                outer.arcMoveTo(x, y, d, d, rangeAngleStart 
+                                                + rangeAngleSpan)
+                                outer.arcTo(x, y, d, d, rangeAngleStart 
+                                            + rangeAngleSpan, -rangeAngleSpan)
+                            else:
+                                # Color scale orientation changed if flipped.
+                                outer.arcMoveTo(x, -y - d, d, d, rangeAngleStart 
+                                                + rangeAngleSpan)
+                                outer.arcTo(x, -y - d, d, d, rangeAngleStart 
+                                            + rangeAngleSpan, -rangeAngleSpan)
+                            qpp.connectPath(outer)
+                            qpp.closeSubpath()
+                            painter.drawPath(qpp)
+                            painter.setTransform(transform)
+
+                            rangeValueStart = breakpoint
 
-            painter.translate(center)
+            painter.resetTransform()
+
+        def drawScaleMarkers(self):
+            painter.setPen(QtGui.QPen(self.palette().color(QtGui.QPalette.Text), 
+                           1))
 
-            painter.rotate(self.m_minimum%ceil(float(majorStep)/float(minorSteps))/float(valueSpan)*angleSpan-angleStart)
+            # Only draw if the scale is visible and enough space for tick marks.
+            if self.m_scaleVisible and majorStep != 0:
 
-            offsetCount = (minorSteps-ceil(self.m_minimum%majorStep)/float(majorStep)*minorSteps)%minorSteps
+                # Rotate the painter to accomodate for vertical scale.
+                if vertical:
+                    if not self.vertically_flipped:
+                        painter.rotate(90)
+                        painter.translate(0, -hWidget + wLabel / 4.0)
+                    else:
+                        painter.rotate(90) 
+                        painter.translate(0, -self.width())
+                        painter.translate(0, -(-hWidget + wLabel / 4.0)) 
+
+                # Account for flipping vertical scales.
+                if vertical and self.vertically_flipped:
+                    painter.translate(center.x(), -center.y())
```

Comment:
This is really kind of general.  Can't we just calculate the correct center in one place instead of tweaking it all over?

Reply:
It could be done that way. I spent a lot of time trying to get the correct orientation for a bunch of things. I would reckon that through trying to find a way to orient things correctly, I did directly change the center value by having the y parameter to be negated. Turns out, I needed to do more than that. I've already figured out that correct configuration (and it was a pain), so I think finding an alternate way to arrange things shouldn't be too high on our priority list as the code works for now.

Reply:
Yep, as I mentioned somewhere else, code like this takes a few passes of pure refactoring to get to the point it's decent.  Having added the functionality is good for now.